bp_fe/src/include/bp_fe_pc_gen.vh

```diff
@@ -0,0 +1,208 @@
+/*
+ * pc_gen.vh
+ *
+ * pc_gen.vh provides the structs for PC generation logics. PC generation use
+ * previous PCs to determine the next PC. PC generation also inspects instrcutions
+ * and flag RVC if the instruction is compressed.  
+ * 
+ * PC generation consists of instruction scanning and Branch Prediction (BP). The
+ * instruction scanning first scans the control flow instructions, to determines
+ * whether the instruction is compressed or not. If the instruction is compressed,
+ * the pc_gen will flag the categories of the instruction to other modules in
+ * backend.  BP uses the previous PC to look up the corresponding entry in the
+ * Branch History Table (BHT).  If the BHT entry predicts taken, BP find the
+ * correpsonding entry in the Branch Target Buffer (BTB) as the next PC.  In the
+ * case of call-ret pairs, BP pushes the call target to Return Address Stack
+ * (RAS) and pops the address in the ret instruction.
+*/
+
+`ifndef BP_FE_PC_GEN_VH
+`define BP_FE_PC_GEN_VH
+
+`ifndef BSG_DEFINES_V
+`define BSG_DEFINES_V
+`include "bsg_defines.v"
+`endif
+
+`ifndef BP_COMMON_FE_BE_IF_VH
+`define BP_COMMON_FE_BE_IF_VH
+`include "bp_common_fe_be_if.vh"
+`endif
+
+// import bp_common_pkg::*;
+import pc_gen_pkg::*;
+
+typedef enum  {
+    e_state_reset               =0
+    ,e_state_first_reset_icache =1
+    ,e_state_first_start_icache =2
+    ,e_state_pc_redirect        =3
+    ,e_state_after_pc_redirect  =4
+    ,e_state_interrupt          =5
+    ,e_state_icache_miss        =6
+    ,e_state_after_icache_miss  =7
+    ,e_state_attaboy            =8
+    ,e_state_no_cmd             =9
+    ,e_state_not_ready          =10
+    ,e_state_always_not_ready   =11
+    ,e_state_invalid_cmd        =12
+} pc_gen_state_e;
+
+// this part is modified to accomodate the instr output
+`define declare_bp_fe_pc_gen_queue_s                    \
+    typedef struct packed {                             \
+        bp_fe_queue_type_e    msg_type;                 \
+        logic [`bp_fe_instr_scan_width-1:0] scan_instr; \
+        union packed {                                  \
+            bp_fe_fetch_s        fetch;                 \
+            bp_fe_exception_s    exception;             \
+        } msg;                                          \
+    } bp_fe_pc_gen_queue_s
+
+`define bp_fe_pc_gen_queue_width(vaddr_width_p,branch_metadata_fwd_width_p) \
+     (`bp_fe_queue_width(vaddr_width_p,branch_metadata_fwd_width_p)+`bp_fe_instr_scan_width)
+
+
```

Comment:
Do we need different structs for these signals? If the PC gen only cares about PC redirection, it might be clearer to just send a redirection packet valid or attaboy valid signal.

Additionally, the compressed logic struct needs some work.  Comment it out?

Reply:
Right, command_queue_opcode can be decoded in fe_top rather than in pc_gen. 
I will change it to just one struct and add redirection_/attaboy_valid signals to it. 

Ok, commented out.